### ISSUE TITLE
source-postgres: On very large tables **Initial** incremental sync is taking too long and times out before starting to emit records

### ISSUE THREAD

**Issue by rodireich** [airbytehq/airbyte#25368]
## Tell us about the problem you're trying to solve
When a table of sizes of dozens of Gigabytes are incrementally synced for the first time, the initial query is ordering by its cursor field. 
This `ORDER BY` is a very lengthy process.
If an index exists on the cursor table, postgres' query planner will utilize this index to do the sort.
But from a certain point when table gets bigger query planner is no longer utilizing this index even if exists and instead opt for doing a lengthy in memory sort.
This causes a large delay in getting the initial records and as a result platform times out waiting for source to start emitting records and sync fails

## Describe the solution you’d like
By changing the **initial** incremental sync query to also include ranges `WHERE cursor > min_val AND cursor < max_val` we can reduce the time it takes to process each chuck of as postgres will be able to use an index to retrieve just the desired range and won't have to sort an entire table upfront.

## Describe the alternative you’ve considered or used
Making changes to db settings such as changing `random_page_cost` or `enable_seqscan` would force query planner to use an index if exist which may solve the initial problem of sync timing out.
But that requires a DB wide change which may have unexpected adverse effects and may make the overall sync go slower.

## Additional context
ref https://github.com/airbytehq/oncall/issues/1819

`ORDER BY` added to initial incremental sync as part of https://github.com/airbytehq/airbyte/pull/24820,
In order to solve cases of failing initial incremental sync which were causing a loss of records.

**Comment by rodireich** [comment]
`ORDER BY` added to initial incremental sync as part of https://github.com/airbytehq/airbyte/pull/24820

**Comment by prateekmukhedkar** [comment]
@rodireich curious, how are `min_val` and `max_val` selected when the incremental sync runs for the first time?

**Comment by happythenewsad** [comment]
Are there plans to implement this? My team is impacted by this issue.

**Comment by happythenewsad** [comment]
I attempted a big data (table size > 1TB) migration from Postgres to Bigquery, after patch #24820 

My syncs still failed with:
`message='activity ScheduleToClose timeout', timeoutType=TIMEOUT_TYPE_SCHEDULE_TO_CLOSE`

**Comment by evantahler** [comment]
@happythenewsad can you share your full sync logs?

**Comment by rodireich** [comment]
This was resolved with our implementation of ctid in https://github.com/airbytehq/airbyte-internal-issues/issues/1786.
The work you're referring to @happythenewsad was done in Apr '23.
Are you trying with the latest source-postgres?

**Comment by evantahler** [comment]
Also noting that the `TIMEOUT_TYPE_SCHEDULE_TO_CLOSE` error reflects a 3-day timeout on all syncs that aren't getting data.

**Comment by happythenewsad** [comment]
Unfortunately my latest attempt was in September 2023, and I no longer have those logs.

**Comment by evantahler** [comment]
Ok, I'll close this issue then.  If you have a more recent problem with logs, please open a new issue!
